clusterloader2/framework/util/util.go

```diff
@@ -0,0 +1,201 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"fmt"
+	"time"
+
+	apiv1 "k8s.io/api/core/v1"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/jsonmergepatch"
+	"k8s.io/apimachinery/pkg/util/mergepatch"
+	utilnet "k8s.io/apimachinery/pkg/util/net"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/dynamic"
+	clientset "k8s.io/client-go/kubernetes"
+)
+
+const (
+	// Parameters for retrying with exponential backoff.
+	retryBackoffInitialDuration = 100 * time.Millisecond
+	retryBackoffFactor          = 3
+	retryBackoffJitter          = 0
+	retryBackoffSteps           = 6
+)
+
+var metadataAccessor = meta.NewAccessor()
+
+// RetryWithExponentialBackOff a utility for retrying the given function with exponential backoff.
+func RetryWithExponentialBackOff(fn wait.ConditionFunc) error {
+	backoff := wait.Backoff{
+		Duration: retryBackoffInitialDuration,
+		Factor:   retryBackoffFactor,
+		Jitter:   retryBackoffJitter,
+		Steps:    retryBackoffSteps,
+	}
+	return wait.ExponentialBackoff(backoff, fn)
+}
+
+// IsRetryableAPIError verifies whether the error is retryable.
+func IsRetryableAPIError(err error) bool {
+	// These errors may indicate a transient error that we can retry in tests.
+	if apierrs.IsInternalError(err) || apierrs.IsTimeout(err) || apierrs.IsServerTimeout(err) ||
+		apierrs.IsTooManyRequests(err) || utilnet.IsProbableEOF(err) || utilnet.IsConnectionReset(err) {
+		return true
+	}
+	// If the error sends the Retry-After header, we respect it as an explicit confirmation we should retry.
+	if _, shouldRetry := apierrs.SuggestsClientDelay(err); shouldRetry {
+		return true
+	}
+	return false
+}
+
+// CreateNamespace creates a single namespace with given name.
+func CreateNamespace(c clientset.Interface, namespace string) error {
+	createFunc := func() (bool, error) {
+		_, err := c.CoreV1().Namespaces().Create(&apiv1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: namespace}})
+		if err == nil || apierrs.IsAlreadyExists(err) {
+			return true, nil
+		}
+		if IsRetryableAPIError(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("Failed to create namespace with non-retryable error: %v", err)
+	}
+	return RetryWithExponentialBackOff(createFunc)
+
+}
+
+// DeleteNamespace deletes namespace with given name.
+func DeleteNamespace(c clientset.Interface, namespace string) error {
+	deleteFunc := func() (bool, error) {
+		err := c.CoreV1().Namespaces().Delete(namespace, nil)
```

Comment:
There is significant conde duplication between those methods here. I think we can refactor them to avoid this.
```
func retryableFunc() func() (bool, error) {
  err := baseFunc()
  if err == nil || isAllowedError(err) {
    return true, nil
  }
  if IsRetryableError(err) {
    return false, nil
  }
  return false, err
}
```

and then:
```
func CreateNamespace(c clientset.Interface, namespace string) error {
  createFunc := func() error {
    _, err := c.CoreV1().Namespaces().Create(...)
   return err
  }
  return RetryWithExponentialBackOff(retryableFunc(baseFunc, apierrs.IsAlreadyExists))
```

[Alternatively, you may consider changing the last line into:
err := Retryable...
if err != nil {
  return fmt.Errorf("failed to create namespace: %v", err)
}
return nil

But I don't think it's really needed.

Reply:
done